sklearn/preprocessing/_csr_polynomial_expansion.pyx

```diff
@@ -81,72 +208,52 @@ def _csr_polynomial_expansion(cnp.ndarray[DATA_T, ndim=1] data,
+    # Make the arrays that will form the CSR matrix of the expansion.
+    cdef INDEX_A_t row_i, row_starts, row_ends, i, j, k, i_ptr, j_ptr, k_ptr
 
-    cdef INDEX_T total_nnz = 0, row_i, nnz
+    cdef INDEX_B_t expanded_index=0, num_cols_in_row, col
 
-    # Count how many nonzero elements the expanded matrix will contain.
+    result_indptr[0] = indptr[0]
     for row_i in range(indptr.shape[0]-1):
-        # nnz is the number of nonzero elements in this row.
-        nnz = indptr[row_i + 1] - indptr[row_i]
-        if degree == 2:
-            total_nnz += (nnz ** 2 + nnz) / 2 - interaction_only * nnz
-        else:
-            total_nnz += ((nnz ** 3 + 3 * nnz ** 2 + 2 * nnz) / 6
-                          - interaction_only * nnz ** 2)
-
-    # Make the arrays that will form the CSR matrix of the expansion.
-    cdef cnp.ndarray[DATA_T, ndim=1] expanded_data = cnp.ndarray(
-        shape=total_nnz, dtype=data.dtype)
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indices = cnp.ndarray(
-        shape=total_nnz, dtype=indices.dtype)
-    cdef INDEX_T num_rows = indptr.shape[0] - 1
-    cdef cnp.ndarray[INDEX_T, ndim=1] expanded_indptr = cnp.ndarray(
-        shape=num_rows + 1, dtype=indptr.dtype)
-
-    cdef INDEX_T expanded_index = 0, row_starts, row_ends, i, j, k, \
-                 i_ptr, j_ptr, k_ptr, num_cols_in_row,  \
-                 expanded_column
-
-    with nogil:
-        expanded_indptr[0] = indptr[0]
-        for row_i in range(indptr.shape[0]-1):
-            row_starts = indptr[row_i]
-            row_ends = indptr[row_i + 1]
-            num_cols_in_row = 0
-            for i_ptr in range(row_starts, row_ends):
-                i = indices[i_ptr]
-                for j_ptr in range(i_ptr + interaction_only, row_ends):
-                    j = indices[j_ptr]
-                    if degree == 2:
-                        col = _deg2_column(d, i, j, interaction_only)
-                        expanded_indices[expanded_index] = col
-                        expanded_data[expanded_index] = (
-                            data[i_ptr] * data[j_ptr])
+        row_starts = indptr[row_i]
+        row_ends = indptr[row_i + 1]
+        num_cols_in_row = 0
+        for i_ptr in range(row_starts, row_ends):
+            i = indices[i_ptr]
+            for j_ptr in range(i_ptr + interaction_only, row_ends):
+                j = indices[j_ptr]
+                if degree == 2:
+                    if max(i, j) > MAX_SAFE_INDEX_DEG2 or i > LONG_MAX // n_features :
+                        # In this case, the Cython implementation
+                        # would result in an integer overflow.
+                        # Here, we take advantage of `PyLong` for arbitrary precision.
+                        with gil:
+                            col = <INDEX_B_t> py_deg2_column(n_features, i, j, interaction_only)
+                    else:
+                        col = <INDEX_B_t> _deg2_column(n_features, i, j, interaction_only)
```

Comment:
Is it possible to move the complexity into `_deg2_column`? In other words, `_deg2_column` will check the condition and go to `py_deg2_column` when it needs to.

My main reason for this is to have a code comment in `_deg2_column` that explains the heuristic for switching. I also prefer to see the constant defined closer to `_deg2_column` so it is easier to reason about.

```python
# explain condition
cdef cnp.int64_t MAX_SAFE_INDEX_DEG2 = <cnp.int64_t> sqrt(LONG_MAX) - 4

cdef inline void _deg2_column(..., INDEX_B_t* col_out) nogil:
    if max(i, y) > MAX_SAFE_INDEX_DEG2:
       with gil:
           col_out[0] = py_deg2_column(...)
    ...
```